clang/lib/Sema/SemaInit.cpp

```diff
@@ -5213,13 +5213,17 @@ static void TryReferenceInitializationCore(Sema &S,
       Sequence.SetOverloadFailure(
                         InitializationSequence::FK_ReferenceInitOverloadFailed,
                                   ConvOvlResult);
-    else if (!InitCategory.isLValue())
-      Sequence.SetFailed(
-          T1Quals.isAddressSpaceSupersetOf(T2Quals)
-              ? InitializationSequence::
-                    FK_NonConstLValueReferenceBindingToTemporary
-              : InitializationSequence::FK_ReferenceInitDropsQualifiers);
-    else {
+    else if (!InitCategory.isLValue()) {
+      if (T1Quals.isAddressSpaceSupersetOf(T2Quals)) {
+        if (!S.getLangOpts().MSVCReferenceBinding ||
+            !S.AllowMSLValueReferenceBinding(T1Quals, T1))
+          Sequence.SetFailed(InitializationSequence::
+                                 FK_NonConstLValueReferenceBindingToTemporary);
```

Comment:
So we were in a branch where this condition holds: `!(T1Quals.hasConst() && !T1Quals.hasVolatile() && T1Quals.isAddressSpaceSupersetOf(T2Quals))`.

Now we enter a branch where `T1Quals.isAddressSpaceSupersetOf(T2Quals)` holds.

Together these two premises imply that `T1Quals.hasConst() && !T1Quals.hasVolatile()` doesn't hold, which is the same as to say `!T1Quals.hasConst() || T1Quals.hasVolatile()` doesn't hold.

This implies that the `AllowMSLValueReferenceBinding` call can only be true iff `Quals.hasVolatile() || QT->isBuiltinType() || QT->isArrayType()` is false, making the `Quals.hasConst()` check in AllowMSLValueReferenceBinding superfluous for this call site.

On the other call site, `T1.isConstQualified() && !T1.isVolatileQualified()` already implies that `AllowMSLValueReferenceBinding` call will return true.

So if I haven't made any mistakes so far (please double check!) you could pretty much perform this rewrite in `AllowMSLValueReferenceBinding`:
```C++
bool Sema::AllowMSLValueReferenceBinding(Qualifiers Quals, QualType QT) {
   return getLangOpts().MSVCReferenceBinding && !(Quals.hasVolatile() || QT->isBuiltinType() || QT->isArrayType());
 }
 ```

Which now justifies this function being a non-static member in Sema, since before it did not need `this` at all.

With this change, it seems the changes needed on the second call site will look much simpler as well, just a one liner:
```C++
if (!isRValRef && (!T1.isConstQualified() || T1.isVolatileQualified() || S.AllowMSLValueReferenceBinding(T1.getQualifiers(), T1))) {
```
But again, please double check!

Reply:
You are correct. I reorder the logic as you have described :).